RES-517: explicit dependency on NuPIC

marionleborgne opened the pull request:
Fixes https://jira.numenta.com/browse/RES-517

rhyolight commented:
I do  not understand this error: https://travis-ci.org/numenta/NAB/builds/226170436

I'm going to re-run the build.

scottpurdy commented:
Hold off on merge, looking into something

scottpurdy commented:
This PR captures the JIRA and what we had discussed correctly. I remembered after looking at the NAB code that we originally intended to not require nupic to be installed. I think it is fine for it to be installed by default, but it shouldn't be required. And we also probably don't want to install nupic if users already have a version installed. I.e. if I do a source install of nupic from master and then install nab, I don't want it to uninstall nupic and install and older version.

So I propose that we extend this PR with the following:

- Mimic the code in nupic for allowing developer installs of nupic.bindings [here](https://github.com/numenta/nupic/blob/23d32924eb9a2392649e3869546699648cddbc49/setup.py#L46-L70). But instead, put this here in NAB and allow developer installs of nupic.
- Update the documentation to communicate that existing nupic installs (if later than 0.5.6) will be left alone but if nupic isn't installed then the target version (0.5.6) will be installed
- Change the example commands to use `pip` instead of directly running `setup.py`
- Also include in the documentation instructions for skipping dependencies (`pip install --no-deps .`) if users want to manage the versions themselves

I'm happy to help with this if it makes sense to you guys (@rhyolight @marionleborgne )

marionleborgne commented:
@scottpurdy This makes sense to me. I'll add these changes.

marionleborgne commented:
@scottpurdy I addressed you feedback in https://github.com/numenta/NAB/pull/301#issuecomment-297548039

[x] Mimic the code in nupic for allowing developer installs of nupic.bindings [here](https://github.com/numenta/nupic/blob/23d32924eb9a2392649e3869546699648cddbc49/setup.py#L46-L70). But instead, put this here in NAB and allow developer installs of nupic.
[x] Update the documentation to communicate that existing nupic installs will be left alone but if nupic isn't installed then the target version will be installed
[x] Change the example commands to use pip instead of directly running setup.py
[x] Also include in the documentation instructions for skipping dependencies (pip install --no-deps .) if users want to manage the versions themselves


Questions:
1. Shouldn't we have `nupic` in requirements.txt instead of `nupic==0.5.6`?
2. I added a check to determine if nupic is installed or not based on [your code example](https://github.com/numenta/nupic/blob/23d32924eb9a2392649e3869546699648cddbc49/setup.py#L46-L70) but in my case this type of logic does find that nupic is installed. I have it installed from source. Any thoughts?

rhyolight commented:
There still needs to be a conditional import for [this](https://github.com/numenta/NAB/blob/1bb58d4a68ff457f4e37a6e31c1ea89a8b08fd41/nab/detectors/numenta/numenta_detector.py#L26) and [this](https://github.com/numenta/NAB/blob/1bb58d4a68ff457f4e37a6e31c1ea89a8b08fd41/nab/detectors/numenta/numentaTM_detector.py#L28). Something like

```python
try:
  from nupic.frameworks.opf.modelfactory import ModelFactory
except:
  from nupic.frameworks.opf.model_factory import ModelFactory
```

marionleborgne commented:
Addressed your feedback @rhyolight

rhyolight commented on nab/detectors/numenta/numentaTM_detector.py:
```diff
@@ -25,7 +25,10 @@
-from nupic.frameworks.opf.modelfactory import ModelFactory
+try:
+  from nupic.frameworks.opf.modelfactory import ModelFactory
+except:  # Try importing it the old way (version < 0.7.0.dev0)
```
`modelfactory` is the old way, `model_factory` is the new way.
Also, [avoid inline comments](http://nupic.docs.numenta.org/contributing/python-style-guide.html#inline-comments).

rhyolight commented:
@marionleborgne back to you ;)

marionleborgne commented:
@rhyolight Gotcha! Fixed :ping_pong:

rhyolight commented:
@scottpurdy Do you want the last review?

scottpurdy commented:
Thanks Marion for putting in the extra effort on this - looks great now!
